dbms/src/Storages/Page/V3/PageDirectory.cpp

```diff
@@ -1623,6 +1720,115 @@ bool PageDirectory<Trait>::tryDumpSnapshot(const ReadLimiterPtr & read_limiter,
     return done_any_io;
 }
 
+template <typename Trait>
+void PageDirectory<Trait>::copyCheckpointInfoFromEdit(PageEntriesEdit & edit)
+{
+    const auto & records = edit.getRecords();
+    if (records.empty())
+        return;
+
+    // Pre-check: All ENTRY edit record must contain remote info.
+    // We do the pre-check before copying any remote info to avoid partial completion.
+    for (const auto & rec : records)
+    {
+        if (rec.type == EditRecordType::VAR_ENTRY)
+            RUNTIME_CHECK(rec.entry.checkpoint_info.has_value());
+    }
+
+    for (const auto & rec : records)
+    {
+        // TODO: Improve from O(nlogn) to O(n).
+
+        typename MVCCMapType::iterator iter;
+        {
+            std::shared_lock read_lock(table_rw_mutex);
+            iter = mvcc_table_directory.find(rec.page_id);
+            if (iter == mvcc_table_directory.end())
+                // There may be obsolete entries deleted.
+                // For example, if there is a `Put 1` with sequence 10, `Del 1` with sequence 11,
+                // and the snapshot sequence is 12, Page with id 1 may be deleted by the gc process.
+                continue;
+        }
+
+        auto & entries = iter->second;
+        entries->copyCheckpointInfoFromEdit(rec);
+    }
+}
+
+template <>
+typename PageDirectory<universal::PageDirectoryTrait>::DumpCheckpointResult
+PageDirectory<universal::PageDirectoryTrait>::dumpIncrementalCheckpoint(const DumpCheckpointOptions & options)
```

Comment:
What about moving this method to `UniversalPageStorage`? This method involves both getting snapshot entries from `PageDirectory` and reading page data from the `BlobStore`

Reply:
Good suggestion! I have moved to UniversalPageStorage.